File: plugin/src/main/com/suhininalex/clones/CloneClass.java
package com.suhininalex.clones;

import com.suhininalex.suffixtree.Edge;
import com.suhininalex.suffixtree.Node;

import java.util.Comparator;

/**
 * Объект теряет целостность при изменении дерева!
 * (Кэш длины может быть неверным, доступ к клонам корректный!)
 */
public class CloneClass {

    private final Node treeNode;
    private int size = -1;
    private final int length;

    public CloneClass(Node treeNode) {
        this.treeNode = treeNode;
        length = computeLengthToRoot(treeNode);
    }

    public Iterable<Clone> getClones(){
        return () -> new CloneIterator(treeNode, getLength());
    }

    public int size(){
        if (size<0) {
            int lastSize = 0;
            for (Clone clone : getClones()){
                lastSize++;
            }
            size = lastSize;
        }
        return size;
    }

    public boolean isEmpty(){
        return !getClones().iterator().hasNext();
    }

    public int getLength() {
        return length;
    }

    public Node getTreeNode(){
        return treeNode;
    }

    @Override
    public String toString() {
        return super.toString();
    }

    public static Comparator<CloneClass> getLengthComparator(){
        return (CloneClass first, CloneClass second)  -> first.getLength() - second.getLength();
    }

    private int computeLengthToRoot(Node node){
        int cloneLength = 0;
        while (node.getParentEdge()!=null){
            Edge edge = node.getParentEdge();
            int edgeLength = edge.getEnd() - edge.getBegin() + 1;
            cloneLength+=edgeLength;
            node = edge.getParent();
        }
        return cloneLength;
    }
}

File: plugin/src/main/com/suhininalex/clones/CloneClass.kt
package com.suhininalex.clones

import com.suhininalex.suffixtree.Edge
import com.suhininalex.suffixtree.Node
import stream
import java.util.stream.Stream

class CloneClass(val treeNode: Node) {

//    val clones: Stream<Clone> get() = treeNode.edges.stream().peek { println("flatmap: ${it.end}") }.flatMap { it.getTerminalsWithOffset() }.map {
//        val (edge, offset) = it
//        val lastElementIndex = edge.end - offset - edge.length
//        val firstElementIndex = lastElementIndex - treeNode.lengthToRoot() + 1
//        println("edge: ${edge.end} ${edge.length} length: $length index:  $firstElementIndex $lastElementIndex $offset")
//        return@map Clone(edge.getFromSequence(firstElementIndex), edge.getFromSequence(lastElementIndex))
//    }

    val clones: Stream<Clone> get() = CloneIterator(this).stream()

    val size by lazy {
        clones.count()
    }

    val length = treeNode.lengthToRoot()

    val isEmpty = clones.isEmpty()

    companion object {
        val lengthComparator = comparator { first: CloneClass, second: CloneClass -> first.length - second.length }
    }

    fun Edge.getTerminalsWithOffset() =
            Pair(this, 0).leafTraverse ({it.first.terminal==null}) {
                val offset = it.first.length + it.second
                println("$offset")
                it.first.terminal.edges.stream().map { Pair(it, offset) }
            }

    private fun Edge.getFromSequence(pos: Int) = sequence[pos] as Token
}